README.md
# node-red-contrib-tellstick

This is a full featured [Node-RED](http://nodered.org/) module to communicate with a [Tellstick](http://www.telldus.se/products/tellstick) or [Tellstick Duo](http://www.telldus.se/products/tellstick_duo).

* Send data to a Tellstick, e.g. turning a lamp on (Available on both Tellstick and Tellstick Duo).
* Recieve data from a (Tellstick compatible) transmitter, e.g a remote control or a temperature sensor (Only available on Tellstick Duo).
* Manage devices, e.g. add and "teach" a new wall plug

## Table of contents
* [Status](#status)
* [Issues](#issues)
* [Looking for older module?](#looking-for-older-module)
* [Example flow](#example-flow)
* [Installation](#installation)
* [Usage, In node](#usage-in-node)
* [Usage, Out node](#usage-out-node)
* [Tips and tricks](#tips-and-tricks)
* [Advanced](#advanced)
* [About the Tellstick](#about-the-tellstick)
* [Developing](#developing)

## Status
What? | Status | What? | Status
------- | ------ | ------- | ------
Codacy | [![Codacy Badge](https://www.codacy.com/project/badge/b2291afcead447048742a9cef0cdd347)](https://www.codacy.com/public/emiloberg/node-red-contrib-tellstick) | Licence | [![Licence](https://img.shields.io/npm/l/node-red-contrib-tellstick.svg)](https://github.com/emiloberg/node-red-contrib-tellstick/blob/master/LICENSE)
Issues | [![Issues](https://img.shields.io/github/issues/emiloberg/node-red-contrib-tellstick.svg)](https://github.com/emiloberg/node-red-contrib-tellstick/issues) | Tag |  [![Tag](https://img.shields.io/github/tag/emiloberg/node-red-contrib-tellstick.svg)](https://github.com/emiloberg/node-red-contrib-tellstick/tags)
GitHub Version | [![GitHub version](https://badge.fury.io/gh/emiloberg%2Fnode-red-contrib-tellstick.svg)](http://badge.fury.io/gh/emiloberg%2Fnode-red-contrib-tellstick) | GitHub Forks | [![Forks](https://img.shields.io/github/forks/emiloberg/node-red-contrib-tellstick.svg)](https://github.com/emiloberg/node-red-contrib-tellstick/network)
NPM Version | [![npm version](https://badge.fury.io/js/node-red-contrib-tellstick.svg)](http://badge.fury.io/js/node-red-contrib-tellstick) | GitHub Followers | [![Followers](https://img.shields.io/github/followers/emiloberg.svg)](https://github.com/emiloberg/followers)
Dependencies | ![Dependencies](https://david-dm.org/emiloberg/node-red-contrib-tellstick.svg)

## Issues
This is still getting developed. Feel free to [add issues, questions or feature requests](https://github.com/emiloberg/node-red-contrib-tellstick/issues).

## Looking for older module?
The NPM namespace `node-red-contrib-tellstick` previously (up until 13 april 2015) held a completely different, simple, module which supported receiving data from Tellstick Duo.

[@japikas](https://github.com/japikas) gracefully transferred ownership of the namespace to this module (which also supports receiving data, plus much more). Are you looking for the old module? Please find it at [https://github.com/japikas/node-red-contrib-tellstick](https://github.com/japikas/node-red-contrib-tellstick). 

## Example flow
Light blue nodes are tellstick nodes.

[![Screenshot - Sample Flows](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-sample-flows.png)](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-sample-flows.png)

## Installation

### 0. Node version
Find out if you're running Node 0.10/0.11 or a newer version. Get your current Node version by running:

```
node -v
```

### 1. Install telldus-core and development libraries

#### Windows and Mac OS X
Make sure you have Telldus Center installed. If you own a Telldus, you probably have this already. If not, [download and install from the Telldus webpage](http://download.telldus.se/TellStick/Software/TelldusCenter/). That's it, go to step 2 below.

#### Ubuntu/Debian/Raspbian (Raspberry Pi)
**Note:** If you previously have installed drivers for the Tellstick, there's a possibility that you've installed `telldus-core` but are missing the `libtelldus-core-dev` which is needed for this module to work on Ubuntu/Debian/Raspbian.

```
#
# Make Telldus available to the distribution
#
sudo nano /etc/apt/sources.list
# add the following line (without the #):
# deb http://download.telldus.com/debian/ stable main

#
# Download and install Telldus repository key
#
wget -q http://download.telldus.se/debian/telldus-public.key -O- | sudo apt-key add -

#
# Update the repository
#
sudo apt-get update

#
# Install telldus-core AND libtelldus-core-dev
#
sudo apt-get install telldus-core libtelldus-core-dev
```

#### Other Linux

Install from source by following this [excellent guide from Lasse](https://lassesunix.wordpress.com/2013/08/12/installing-telldus-core-on-raspberry-pi/).

### 2. Install module(s)

For some reason, the underlying telldus library does not have node-gyp as a dependency (even though it actually needs it). Therefor install by `cd`-ing to your Node-RED directory (`$HOME/.node-red`) and run these 3 commands in sequence.

#### 1)
```
npm install node-gyp
```

#### 2)
**Run one of these commands:**

If you're running Node 0.12 or above, run:

```
npm install telldus
```

or, if you're running a legacy version of Node (0.10/0.11), run:

```
npm install telldus-legacy
```

#### 3)
```
npm install node-red-contrib-tellstick
```


## Usage

This module consist of 2 nodes, one for listening to incoming data (remotes and sensors) and one for transmitting data (e.g telling a lamp to turn off).

## Usage, _In_ node
This module will listen to all incoming data and if that data matches the rule set in the _in_ node, the node will be triggered and pass the complete incoming message object forward.

E.g. you can set an input node to listen to all remote which sends `house`=_15414550_ and `group`=0. This will fire the node on all button press' on that remote control. If you want to qualify it you may adda a `unit`=10 to the rule which will make it only fire on button _10_. Or qualify it even more by adding the rule `method`=turnoff to only make it trigger on the _off_ button (and not the _on_ button).

You can easily make your rules by placing a Tellstick in node on your Sheet, clicking it and select "Add new tellstick-input". A dialog will open which will show _all_ incoming data. Just press a button on the remote (or wait for your sensor to transmit a signal) and it will show up on the dialog. From there you may can the _learn_ button which will populate the rule fields.

If you want to create an In node which passes on _all_ incoming data, just create a config with all fields left blank (meaning: no rules).

[![Screenshot of input configuration](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-in-config.png)](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-in-config.png)

If a in node is triggered, it will output the complete data message from the transmitter. Check the message out by wiring the input node to a debug node (set to display _full message_).

Example of outputted data. However, note that no transmitter sends _all_ fields. All transmitters have fields like `class` and `protocol` but e.g. _codeswitch_ type transmitters only transmit `code` and not `house` or `unit` where for a _selflearning_ transmitter it's the opposite.
 
```
{
    "class": "command",
    	// Could be 'command' for remote or 'sensor' for sensor
    "protocol": "sartano",
    "group": 0,
    "house": 12345678,
    "method": "turnon",
	    // Remotes usually send "turnon" for "on" and "turnoff" for "off" button
    "model": "codeswitch",
    "unit": 1,        
    "code": 1011011011,
    "id": 123,
    "additionalDataFields": 22.1
    	// A sensor could send "temp: 22.1" for temperature reading,
    	// or "humidity: 57" for humidity

}
```

Example, pressing the first off button on a Nexa handheld remote:

```
{
	class: 'command',
	protocol: 'everflourish',
	model: 'selflearning',
	house: 34937,
		// Built in identifyer for this particular remote
	unit: 1,
		// Number 1 button (this remote have 4 pairs of on/off buttons)
	method: 'turnoff'
		// Pressed the "off" button. On button would give "turnon"
}
```

Example Proove TSS320 Temperature/Hygrometer sensor. Sends a signal every minute.

```
{ 
	class: 'sensor',
	protocol: 'fineoffset',
	id: 183,
	model: 'temperaturehumidity',
	humidity: 37,
	temp: 22
 }
```

## Usage, _Out_ node

The out node will send a signal telling your heater to turn off or your lamp to dim to 45%. 

Example of an out node, dimming the device "Red lightsaber" to 165 (65%):

[![Screenshot of output configuration](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-out.png)](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-out.png)

### Passing data to the out node.

There are 3 ways of passing data to an out node:

* Configure the `device`, `method` (turn on/turn off/dim) and `dimlevel` (0-255) if a dim capable device on the output node itself (as done in the screenshot above)
* Send the data in the `msg` sent to the out node.
* A combination of above. E.g set the device to _Red lightsaber_ in the configuration dialog but send `method: turnon` or `method: turnoff` in the `msg` to the out node. Whatever is sent in the msg will always override the settings set in the node.

```
msg: {
	device: 1,
		// Integer, device ID. 
		// The Device IDs are displayed in the "Configure devices"
		// dialog (se further down in this README).
	method: 'turnon',
		// Valid methods: turnoff, turnon, dim, bell, learn
		// if dim is supplied so must dimlevel be.
		// method may also be an integer:
		//		0 (for turnoff)
		//		1 (for turnon)
		//		2 (for dim)
		//		3 (for bell)
		//		4 (for learn)
	dimlevel: 100
		// Only taken into account if method is "dim".
		// Valid input: 0-255, where 255 is 100% and 0 is off.
}
```

### Configure devices

The way the Tellstick works you have to configure your devices. On Windows and Mac OS X you do this in Telldus Center, on Linux you do this in `/etc/tellstick.conf`. However, this is a full featured module to handle Tellstick devices and you may add/update/delete devices straight from Node-RED with this module. Any changes you do from Node-RED will be saved (e.g. written to `/etc/tellstick.conf`).

To configure devices. Place an Out node on a Sheet, click it and click the big blue "Configure/Add devices". A new dialog will open showing all configured devices. From here you can also test your devices but turning them on/off or dimming them. Or you may send the "learn" command to them.

[![Screenshot of device list](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-devicelist.png)](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-devicelist.png)

Clicking "Configure" or "Add new device" will give you a dialog to add/configure/remove a device.

[![Screenshot of input configuration](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-update-device.png)](https://raw.githubusercontent.com/emiloberg/node-red-contrib-tellstick/master/docs/screenshot-update-device.png)

## Tips and tricks
### Converting "on" to "dim to n"
Sending the "turnon" command to a dim capable device will turn it on to the same dim level as you last set it to. If you have a regular on/off remote and want to convert a "turnon" command to a "dim to maximum" you can easily drop a function node between your Tellstick In node and your Tellstick Out node converting the message:

```
if (msg.method === 'turnon') {
	msg.dimlevel = 255;
	msg.method = 'dim'
}

return msg;
```

`dimlevel` goes from 0 to 255, so 255 will dim the device to 100%.


## Advanced

### Sending same command multiple times

The _tellstick-out_ node sends the same command multiple times to ensure that at least one of them reaches the receiver.

As the Tellstick just sends radio signals without any way of knowing if the transmission reched its goal, sometimes signals gets lost in space.

To mitigate this. The _tellstick-out_ sends the same command multiple times.

By default, it sends the same command 5 times with an interval of 500 milliseconds (0.5 seconds) between each command. You can tweak these settings but editing the `settings.js` file in your Node-RED directory (the path to the settings file is printed in the console when you're starting Node-RED).

The new settings are:

```
functionGlobalContext: {
	repeatSendTimes: 5,
	repeatSendInterval: 500
}
```

#### Possible caveat

Imagine that you have a flow that will turn on a lamp (command A) and directly thereafter turn off that lamp (command B). 

Then the command A will turn the lamp on, then command B will turn the lamp off. Then the command A will emit it's second signal, turning the lamp on again. Command B will emit it's second signal, turning the lamp on - and so on. 

If you have a flow as such, then set `repeatSendTimes` to `1`. However, this should really be a corner case. 

### Debounce/Throttle Data
A radio transmitter will often send the same command multiple times in a short rapid burst to ensure that at least one of them is received. Most 433 Mhz transmitters send the same command 5-6 times.

To make sure it only gets picked up once (even if all 6 of those transmissions makes it through to the Tellstick) we need "debounce" the data. If the same command gets received within a certain time frame, we treat it as one (1) command. By default this time is 500ms.

When _transmitting_ data, the Tellstick behaves in the same way, it will send the same command about 6 times. However, this means that if we're sending two different commands, the Tellstick might start sending the second command before it's done with the first command. This makes only the first receiver pick up a clean signal and do what it's supposed to do (like turn on a lamp). The second receiver will get scrambled data and probably just ignore it. 

To solve this we've to make sure the Tellstick is only sending one command at a time. The default time between 2 commands is 900ms (discovered by the nobel art of trial and error).

Unlikely but if your input nodes are getting fired twice or if a second receiver aren't getting the signal (e.g. aren't turning on) you may tweak these times but editing the `settings.js` file in your Node-RED directory.

Add the `tellstickInputThrottle` and/or `tellstickOutputThrottle` property to the `functionGlobalContext` object like this:

```
functionGlobalContext: {
	tellstickInputThrottle: 500,
	tellstickOutputThrottle: 900
}
```

## About the Tellstick
The Telldus Tellstick is a small USB connected radio tranciever which plays well on Windows/Mac/Linux (Raspberry) and is used to control a wide range of 433 MHz based plugged in and built in electric switches and dimmers. Further, this module can also listen to _incoming_ Tellstick data (or in other words: it can pick up signals from remote controls).

There are 3 different types of Tellsticks:

* [Tellstick](http://www.telldus.se/products/tellstick) - The simplest device. This device can only transmit radio (meaning you may turn on a lamp from Node-RED)
* [Tellstick Duo](http://www.telldus.se/products/tellstick_duo) - Transmitter and receiver. This is most probably the one you want. With this you can transmit radio as well as receive (meaning Node-RED may listen and act on incoming radio, e.g. from a sensor or remote control)
* [Tellstick Net](http://www.telldus.se/products/tellstick_net) -  This node does _not_ support this device. Ethernet device. It's more or less an all-in-one solution with its own software.

#### Commonly used Tellstick transmitters/receivers 
* [HomeEasy](http://www.homeeasy.eu/)
* Nexa [Swedish](http://www.nexa.se/) | [English](http://www.nexa.se/Home.html) (Resellers include Clas Ohlson in [United Kingdom](http://www.clasohlson.com/uk/), [Sweden](http://www.clasohlson.com/se/), [Norway](http://www.clasohlson.com/no/), [Finland](http://www.clasohlson.com/fi/) and [UAE](http://www.clasohlson.com/ae/)
* [Proove](http://proove.se/) (Swedish)
* [X10](http://en.wikipedia.org/wiki/X10_%28industry_standard%29)
* [Kjell & Company](http://www.kjell.com/) (Swedish)

#### Other supported Tellstick transmitters/receivers 
Anslut, Brennestuhl, Bye Bye Standby, Chacon, CoCo Technologies, Conrad, Ecosavers, Elro, GAO, Goobay, HQ, IKEA, Intertechno, Kappa, KlikAanKlikUit, Otio, Rusta, Sartano, UPM, Waveman.

See [complete list of supported devices](http://telldus.se/products/compability) on Telldus webpage.


## Developing
This modules is quite big and to make it easier to handle, the source code is is broken down into smaller files.

As per Node-RED standard, each module consist of 1 html file and 1 js file. These files lives in the `/tellstick` folder and **gets built** from the `/lib` folder. This is, somewhat incorrectly, named _lib_ because it's a reserved word in the Node-RED world and Node-RED will not try to discover nodes in folders named _lib_. The more correct folder name would be _source_.

A Gulp build script is included. Run `gulp build` to build the html/js files. Run `gulp` to start gulp in developer mode and auto-build the files when the source files (in `/lib`) are changed.

CHANGELOG.md
# node-red-contrib-tellstick changelog

## 2.1.3
Fixed [dependency warning](https://github.com/emiloberg/node-red-contrib-tellstick/issues/8). Thanks [Hexagon](https://github.com/Hexagon).

## 2.1.2

Bumped [telldus](https://github.com/Hexagon/node-telldus) dependency to 1.0.0

## 2.1.1

Fixed missing changelog

## 2.1.0

Added support for Node 0.12+

## 2.0.0

### Possible breaking changes

**The _tellstick-out_ node now sends the same command multiple times** to ensure that at least one of them reaches the receiver.

As the Tellstick just sends radio signals without any way of knowing if the transmission reched its goal, sometimes signals gets lost in space.

To mitigate this. The _tellstick-out_ now sends the same command multiple times.

By default, it sends the same command 5 times with an interval of 500 milliseconds (0.5 seconds) between each command. You can tweak these settings but editing the `settings.js` file in your Node-RED directory (the path to the settings file is printed in the console when you're starting Node-RED).

The new settings are:

```
functionGlobalContext: {
	repeatSendTimes: 5,
	repeatSendInterval: 500
}
```

#### Why could this be breaking?

Imagine that you have a flow that will turn on a lamp (command A) and directly thereafter turn off that lamp (command B). 

Then the command A will turn the lamp on, then command B will turn the lamp off. Then the command A will emit it's second signal, turning the lamp on again. Command B will emit it's second signal, turning the lamp on - and so on.

If you have a flow as such, then set `repeatSendTimes` to `1`. However, this should really be a corner case.

tellstick/tellstick-out.js
'use strict';

var telldus = require('telldus');
var is = require('is_js');
var Settings = require('./lib/settings.js');

/**
 * Normalize and validate an incoming command.
 *
 * Merge any properties coming in the incoming data object with the
 * properties set as settings on the output node.
 *
 * @param incomingMsg
 * @param node
 * @returns deviceMethodObject
 */
function normalizeTellstickCommand(incomingMsg, node) {
	/**
	 * Merge incoming msg with config set on the node.
	 * Incoming msg will take precedence over configuration set
	 * on the node itself.
	 */
	var msg = {
		device: '',
		method: '',
		dimlevel: ''
	};
	Object.keys(msg).forEach(function (key) {
		if (typeof incomingMsg === 'object') {
			if (incomingMsg.hasOwnProperty(key)) {
				msg[key] = incomingMsg[key];
			} else {
				if (node) {
					msg[key] = node[key];
				}
			}
		} else {
			if (node) {
				msg[key] = node[key];
			}
		}
	});
	msg.err = false;
	msg.warn = false;
	msg.errWarnStr = '';


	/**
	 * We allow the incoming msg to be a little bit ambiguous,
	 * so lets normalize it.
	 */
	msg.device = parseInt(msg.device);
	msg.dimlevel = parseInt(msg.dimlevel);
	if (msg.method === 'turnon' || msg.method === '1') {
		msg.method = 1;
	} else if (msg.method === 'turnoff' || msg.method === '0') {
		msg.method = 0;
	} else if (msg.method === 'dim' || msg.method === '2') {
		msg.method = 2;
	} else if (msg.method === 'bell' || msg.method === '3') {
		msg.method = 3;
	} else if (msg.method === 'learn' || msg.method === '4') {
		msg.method = 4;
	} else {
		msg.err = true;
		msg.errWarnStr = "No valid method is supplied. Please select a method in the output node, or send a method in the msg: 'msg: {method: 'turnon'}'. Valid method are 'turnon' (or '1'), 'turnoff' (or '0'), 'dim' (or '2')'";
	}
	/**
	 * Convert 'dim to 0' to 'turnoff'
	 */
	if (msg.method === 2 && msg.dimlevel === 0) {
		msg.method = 0;
	}


	/**
	 * Validate msg
	 */
	if (is.not.integer(msg.device)) {
		msg.err = true;
		msg.errWarnStr = "Value for 'device' is not a number. Please select a device in the output node, or send a device in the msg: 'msg: {device: 1}'";
	}

	if (msg.method === 2) {
		if (is.integer(msg.dimlevel)) {
			if (msg.dimlevel < 0) {
				msg.dimlevel = 0;
				msg.warn = true;
				msg.errWarnStr = "Value for 'device' is not a number. Please select a device in the output node, or send a device in the msg: 'msg: {device: 1}'";
			} else if (msg.dimlevel > 255) {
				msg.dimlevel = 255;
				msg.warn = true;
				msg.errWarnStr = "'dimlevel' is > 255, setting dimlevel to 255 (max value)";
			}
		} else {
			msg.err = true;
			msg.errWarnStr = "'method' is set to 'dim' but no valid 'dimlevel' is supplied. Needs to be a number between 0 and 255.";
		}
	}

	return msg;
}


module.exports = function(RED) {
	var tellstickDevices = require('./lib/tellstickDevices.js');
	var tellstickMethods = require('./lib/tellstickMethods.js');

	/**
	 * If set: Read output throttle time from Node-RED settings
	 * and update our settings with that value.
	 */
	if (RED.settings.functionGlobalContext.hasOwnProperty('tellstickOutputThrottle')) {
		Settings.update('outputThrottle', RED.settings.functionGlobalContext.tellstickOutputThrottle);
	}
	if (RED.settings.functionGlobalContext.hasOwnProperty('repeatSendTimes')) {
		Settings.update('repeatSendTimes', RED.settings.functionGlobalContext.repeatSendTimes);
	}
	if (RED.settings.functionGlobalContext.hasOwnProperty('repeatSendInterval')) {
		Settings.update('repeatSendInterval', RED.settings.functionGlobalContext.repeatSendInterval);
	}

	/**
	 * Create Node
	 * @param n
	 * @constructor
	 */
	function TellstickOutNode(n) {
		RED.nodes.createNode(this, n);
		this.name = n.name;
		this.device = n.device || '';
		this.method = n.method || '';
		this.dimlevel = n.dimlevel || '';
		this.devicefriendlyname = n.devicefriendlyname || '';

		/**
		 * Display Tellstick Status on Node
		 */
		var tellstickEvents = require('./lib/tellstickEvents.js');

		var tellstickStatus = tellstickEvents.getStatus();
		if (tellstickStatus.status === 0) {
			this.status({fill: 'green', shape: 'ring', text: 'connected'});
		} else {
			this.status({fill: 'red', shape: 'ring', text: tellstickStatus.errStr});
		}

		var node = this;


		/**
		 * On Input
		 */
		this.on('input', function(incomingMsg) {

			var msg = normalizeTellstickCommand(incomingMsg, node);

			if (msg.err === true) {
				this.error(msg.errWarnStr);
				return;
			} else if (msg.warn === true) {
				this.warn(msg.errWarnStr);
			}

			for (var i = 0; i < Settings.get('repeatSendTimes'); i++) {
				if (i === 0) {
					tellstickMethods.queue(msg, function (err) {
						if (err) {
							node.error(JSON.stringify(err.message));
						}
					});
				} else {
					setTimeout(function () {
						tellstickMethods.queue(msg);
					}, i * Settings.get('repeatSendInterval'));
				}
			}

		});
	}
	RED.nodes.registerType('tellstick-out', TellstickOutNode);



	/**
	 * Add new device
	 */
	function addUpdateDevice(updateDeviceId, deviceObj, cb) {
		if (is.not.integer(updateDeviceId)) {
			throw new Error('updateDeviceId must be a number');
		}
		var deviceId = updateDeviceId;
		var doneStatus = 'updated';
		if (updateDeviceId === -1) {
			deviceId = telldus.addDeviceSync();
			doneStatus = 'added';
		}
		if (is.under(deviceId, 1)) {
			cb(new Error('Could not add device: ' + telldus.getErrorString(deviceId)));
			return;
		}
		telldus.setName(deviceId, deviceObj.name, function () {
			telldus.setProtocol(deviceId, deviceObj.protocol, function () {
				telldus.setModel(deviceId, deviceObj.model, function() {
					Object.keys(deviceObj.parameters).forEach(function (param) {
						telldus.setDeviceParameterSync(deviceId, param, deviceObj.parameters[param]);
					});
					deviceObj.status = doneStatus;
					cb(null, deviceObj);
				});
			});
		});
	}


	/**
	 * API: Give the client a list of all brands
	 */
	RED.httpAdmin.get('/tellstick/supported-brands', function(req, res) {
		res.writeHead(200, {'Content-Type': 'application/json'});
		res.write(JSON.stringify(tellstickDevices.getBrands()));
		res.end();
	});


	/**
	 * API: Give the client a list of all device types for a given brand
	 */
	RED.httpAdmin.get('/tellstick/supported-devices/:brand', function(req, res) {
		res.writeHead(200, {'Content-Type': 'application/json'});
		res.write(JSON.stringify(tellstickDevices.getModels(req.params.brand)));
		res.end();
	});


	/**
	 * API: Add new device/Update device
	 */
	RED.httpAdmin.post('/tellstick/device', function(req, res) {

		var updateDeviceId = parseInt(req.body.id);
		if (is.not.integer(updateDeviceId)) {
			updateDeviceId = -1;
		}

		addUpdateDevice(updateDeviceId, req.body, function (err, data) {
			if (err) {
				res.writeHead(500, {'Content-Type': 'application/json'});
				res.write(JSON.stringify(err.message));
				res.end();
				return;
			}
			res.writeHead(200, {'Content-Type': 'application/json'});
			res.write(JSON.stringify(data));
			res.end();
		});

	});


	/**
	 * API: Invoke method (on, off, dim, bell, learn)
	 *
	 */
	RED.httpAdmin.get('/tellstick/invoke/:deviceid/:method/:dimlevel?', function(req, res) {
		var incomingMsg = {
			device: req.params.deviceid,
			method: req.params.method,
			dimlevel: req.params.dimlevel || ''
		};

		var out = normalizeTellstickCommand(incomingMsg);

		if (out.err === true) {
			res.writeHead(400, {'Content-Type': 'application/json'});
			res.write(JSON.stringify(out));
			res.end();
			return;
		}

		for (var i = 0; i < Settings.get('repeatSendTimes'); i++) {
			if (i === 0) {
				tellstickMethods.queue(out, function(err, status) {
					if (err) {
						res.writeHead(500, {'Content-Type': 'application/json'});
						out.err = true;
						out.errWarnStr = err.message;
						res.write(JSON.stringify(out));
						res.end();
					} else {
						out.status = status;
						res.writeHead(200, {'Content-Type': 'application/json'});
						res.write(JSON.stringify(out));
						res.end();
					}
				});
			} else {
				setTimeout(function() {
					tellstickMethods.queue(out);
				}, i * Settings.get('repeatSendInterval'));
			}
		}

	});


	/**
	 * API: Give the client a single device and its configured parameters.
	 */
	RED.httpAdmin.get('/tellstick/device/:deviceid', function(req, res) {

		var deviceId = parseInt(req.params.deviceid);

		if (is.not.integer(deviceId)) {
			res.writeHead(400, {'Content-Type': 'application/json'});
			res.write(JSON.stringify({
				errStr: 'Device needs to be a number'
			}));
			res.end();
			return;
		}

		tellstickDevices.getDevices(deviceId, function (errDevices, device) {
			if (errDevices) {
				res.writeHead(500, {'Content-Type': 'application/json'});
				res.write(JSON.stringify(errDevices));
				res.end();
			} else {
				tellstickDevices.getParametersValues(deviceId, function (errParameters, parameters) {
					if (errParameters) {
						res.writeHead(500, {'Content-Type': 'application/json'});
						res.write(JSON.stringify(errParameters));
						res.end();
					} else {
						device.parameters = parameters;
						device.brand = tellstickDevices.getBrandFromModel(device.model);
						res.writeHead(200, {'Content-Type': 'application/json'});
						res.write(JSON.stringify(device));
						res.end();
					}
				});
			}
		});
	});


	/**
	 * API: Give the client a list of all configured devices
	 */
	RED.httpAdmin.get('/tellstick/device', function(req, res) {
		tellstickDevices.getDevices(function (err, data) {
			if (err) {
				res.writeHead(500, {'Content-Type': 'application/json'});
				res.write(JSON.stringify(err));
				res.end();
				return;
			}
			res.writeHead(200, {'Content-Type': 'application/json'});
			res.write(JSON.stringify(data));
			res.end();
		});
	});


	/**
	 * API: Delete Device
	 */
	RED.httpAdmin.delete('/tellstick/device/:id', function(req, res) {
		var deviceId = parseInt(req.params.id);

		if (is.not.integer(deviceId)) {
			res.writeHead(400, {'Content-Type': 'application/json'});
			res.write(JSON.stringify({
				errStr: 'deviceId needs to be a number'
			}));
			res.end();
			return;
		}

		telldus.removeDevice(deviceId, function(err) {
			if (err) {
				res.writeHead(404, {'Content-Type': 'application/json'});
				res.write(JSON.stringify({
					errStr: err.message
				}));
				res.end();
				return;
			}
			res.writeHead(200, {'Content-Type': 'application/json'});
			res.write(JSON.stringify({
				id: deviceId,
				status: 'deleted'
			}));
			res.end();
		});
	});


};
